docs/MV3.md

```diff
@@ -0,0 +1,114 @@
+
+![MV3 Working Diagram](assets/mv3-working.png)
+
+The process is asynchronous, the browser allows "observation" of requests to companion, which asynchronously determines if the given request is serviceable by IPFS and then dynamically introduces a rule for the browser to perform redirects to the local gateway. A sample rule looks something like:
+
+```js
+{
+  id,                                    // random id for reference
+  priority: 1,                           // priority of the rule
+  action: {
+    type: 'redirect',
+    redirect: { '<regex substitution>' } // redirect to gateway
+  },
+  condition: {
+    '<regex filter>',                    // filter to match requests
+    '<filtered domains list>,
+    resourceTypes: [
+      'csp_report',
+      'font',
+      'image',
+      'main_frame',
+      'media',
+      'object',
+      'other',
+      'ping',
+      'script',
+      'stylesheet',
+      'sub_frame',
+      'webbundle',
+      'websocket',
+      'webtransport',
+      'xmlhttprequest'
+    ]
+  }
+}
+```
+
+The API only allows for 5000 such rules to exist at the time of writing this, so that value can be exhausted pretty quickly. So instead of adding one rule per redirect, we can dynamically generate a single regex pattern to handle multiple redirections.
+
+For example
+
+```
+https://ipfs.io/ipns/en.wikipedia-on-ipfs.org -> http://localhost:8080/ipns/en.wikipedia-on-ipfs.org
+```
+
+can be represented as
+
+```
+https://ipfs.io/ipns/(.*) -> http://localhost:8080/ipns/$1
+```
+
+That's just an example, but in reality it looks more like (which covers edge cases like URL endings):
+
+```js
+{
+  "action": {
+    "redirect": {
+      "regexSubstitution": "http://localhost:8080\\1"
+    },
+    "type": "redirect"
+  },
+  "condition": {
+    "excludedInitiatorDomains": [],
+    "regexFilter": "^https?\\:\\/\\/ipfs\\.io((?:[^\\.]|$).*)$",
+    "resourceTypes": [
+      "csp_report",
+      "font",
+      "image",
+      "main_frame",
+      "media",
+      "object",
+      "other",
+      "ping",
+      "script",
+      "stylesheet",
+      "sub_frame",
+      "webbundle",
+      "websocket",
+      "webtransport",
+      "xmlhttprequest"
+    ]
+  },
+  "id": 1234,
+  "priority": 1
+}
+```
+
+This is a single rule that covers all the redirects from `ipfs.io` to `localhost:8080`. This means that we can save a lot of rules and gain more coverage of serviceable URLs. This is the approach we have taken in MV3.
+
+## Gotchas
+
+Since the process is asynchronous, there are a few things that we need to keep in mind:
+
+- There will always be scenarios, where the redirection to public URL happens before companion determines if the URL is serviceable (consider ipns/fqdn resolutions.) To tackle this a refresh mechanism has been put in place that refreshes the page after a few seconds if the URL is serviceable. This is not ideal, but it is the best we can do for now.
```

Comment:
> There will always be scenarios, where the redirection to public URL happens before companion determines if the URL is serviceable

Adding the word "redirection" seems odd here.
Maybe "where the servicing of public URLs happen..."